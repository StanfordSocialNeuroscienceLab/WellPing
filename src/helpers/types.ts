import { QuestionType } from "./helpers";

export type StreamName = string;

export type QuestionId = string;

export interface Question {
  id: QuestionId;
  type: QuestionType;
  question: string;
  next?: QuestionId | null;
}

export interface SliderQuestion extends Question {
  type: QuestionType.Slider;
  slider: [string, string]; // [left, right]
  defaultValue?: number;
  defaultValueFromQuestionId?: QuestionId;
}

export type Choice = {
  key: string;
  value: string;
};

export interface ChoicesQuestion extends Question {
  choices: Choice[];
  specialCasesStartId?: {
    [questionId: string /* actually QuestionId */]: QuestionId;
    _pna?: QuestionId;
  };
  randomizeChoicesOrder?: boolean;
  randomizeExceptForChoiceIds?: string[];
}

export interface ChoicesWithSingleAnswerQuestion extends ChoicesQuestion {
  type: QuestionType.ChoicesWithSingleAnswer;
}

export interface ChoicesWithMultipleAnswersQuestion extends ChoicesQuestion {
  type: QuestionType.ChoicesWithMultipleAnswers;
}

export interface YesNoQuestion extends Question {
  type: QuestionType.YesNo;
  branchStartId?: {
    yes?: QuestionId;
    no?: QuestionId;
  };
  // Currently only `yes` is supported and also can only followup after 3 days and 7 days with the same stream.
  addFollowupStream?: {
    yes?: StreamName;
    // TODO: no?: StreamName;
  };
}

export interface MultipleTextQuestion extends Question {
  // `id` will store the number of text fields answered.
  type: QuestionType.MultipleText;
  eachId: QuestionId;
  placeholder?: string;
  choices?: "NAMES" | Choice[];
  forceChoice?: boolean;
  max: number;
  maxMinus?: QuestionId; // The max number of text field will be `max` minus the number of text the participant entered in `maxMinus` question.
  repeatedItemStartId?: QuestionId;
  fallbackItemStartId?: QuestionId; // This is used when the user does not enter any name or select prefer not to answer. Note that this has to exists somewhere else. If it is `null`, we will go to `next` directly.
  indexName: string;
  variableName: string;
}

export interface HowLongAgoQuestion extends Question {
  type: QuestionType.HowLongAgo;
}

export interface BranchQuestion extends Question {
  // This is not actually a question (it will not be displayed to the user)
  type: QuestionType.Branch;
  condition: {
    questionId: QuestionId;
    questionType:
      | QuestionType.MultipleText
      | QuestionType.ChoicesWithSingleAnswer;
    compare: "equal";
    target: number | string;
  };
  branchStartId: {
    true?: QuestionId;
    false?: QuestionId;
  };
}

export interface BranchWithRelativeComparisonQuestion extends Question {
  // This is not actually a question (it will not be displayed to the user)
  type: QuestionType.BranchWithRelativeComparison;
  branchStartId: {
    [questionId: string /* actually QuestionId */]: QuestionId;
  };
}

export interface QuestionsList {
  [id: string]: Question;
}

export type Streams = {
  [stream: string /* actually StreamName */]: QuestionsList;
};

export type StreamStartingQuestionIds = {
  [stream: string /* actually StreamName */]: QuestionId;
};

export type StudyID = string;

export type StudyInfo = {
  id: StudyID;
  consentFormUrl: string;
  contactEmail?: string;
  weekStartsOn: 0 | 1 | 2 | 3 | 4 | 5 | 6; // 0 is Sunday
  startDate: Date; // First survey will be sent after this time.
  endDate: Date; // Last survey will be sent before this time.
  frequency: {
    hoursEveryday: number[];

    // Randomly add a number of minutes between `min` to `max` minutes (inclusive) to the notification time.
    randomMinuteAddition: {
      min: number;
      max: number;
    };
  };
  streamsOrder: {
    // Notice the numbering here is not affected by the `weekStartsOn` option above.
    // The number of elements each day should be equal to the number of elements in `frequency.hoursEveryday` above.
    0: StreamName[]; // Sunday
    1: StreamName[]; // Monday
    2: StreamName[]; // Tuesday
    3: StreamName[]; // Wednesday
    4: StreamName[]; // Thursday
    5: StreamName[]; // Friday
    6: StreamName[]; // Saturday
  };

  // A stream that the user will see by default if there is any error in `streamsOrder`.
  streamInCaseOfError: StreamName;

  // A list of streams that will not be replaced by followup streams
  // (when a YesNo question's `addFollowupStream` is set).
  streamsNotReplacedByFollowupStream?: StreamName[];

  notificationContent: {
    default: {
      title: string;
      body: string;
    };
    // We will use `bonus` content for the notification until the number of survey completed that week is greater than `bonus.numberOfCompletionEachWeek - 1`.
    // `#n_ping#` will turn into "5 pings" or "1 ping".
    bonus: {
      title: string;
      body: string;
      numberOfCompletionEachWeek: number;
    };
  };
  serverURL: string;
};

export type SurveyFile = {
  studyInfo: StudyInfo;
  meta: {
    startingQuestionIds: StreamStartingQuestionIds;
  };
  streams: Streams;
};

export type Names = string[];
