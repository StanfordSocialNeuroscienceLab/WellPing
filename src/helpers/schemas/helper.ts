export const idRegexCheck = (val: string) => /^\w+$/.test(val);
export const idRegexErrorMessage = (name: string) =>
  `${name} can only include letters, numbers, and "_".`;
