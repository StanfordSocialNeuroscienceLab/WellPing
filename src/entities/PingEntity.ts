import { BaseEntity, Entity, PrimaryColumn, Column } from "typeorm";

import { StreamName } from "../helpers/types";

@Entity()
export default class PingEntity extends BaseEntity {
  @PrimaryColumn()
  id: string;

  @Column()
  notificationTime: Date;

  @Column()
  startTime: Date;

  @Column({ nullable: true })
  endTime?: Date;

  @Column()
  tzOffset: number;

  @Column({ type: "varchar" })
  streamName: StreamName;
}
