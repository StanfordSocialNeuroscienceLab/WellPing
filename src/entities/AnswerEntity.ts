import {
  BaseEntity,
  Entity,
  ChildEntity,
  Column,
  ManyToOne,
  RelationId,
  TableInheritance,
} from "typeorm";

import {
  AnswerData,
  SliderAnswerData,
  ChoicesWithSingleAnswerAnswerData,
  ChoicesWithMultipleAnswersAnswerData,
  YesNoAnswerData,
  MultipleTextAnswerData,
  HowLongAgoAnswerData,
} from "../helpers/answerTypes";
import { QuestionType } from "../helpers/helpers";
import { QuestionId, QuestionTypeType } from "../helpers/types";
import { PingEntity } from "./PingEntity";

@Entity("answer")
@TableInheritance({
  column: { type: "simple-enum", name: "questionType", enum: QuestionType },
})
export abstract class AnswerEntity extends BaseEntity {
  // https://github.com/typeorm/typeorm/issues/4190
  // TODO: remove me once typeorm is updated to a newer version
  @ManyToOne("PingEntity", "answers", { primary: true })
  ping: PingEntity;

  @RelationId((answer: AnswerEntity) => answer.ping)
  pingId: number;

  @Column({ type: "varchar", primary: true })
  questionId: QuestionId;

  /**
   * MARK: WHY_PNA_TRUE_OR_NULL
   *
   * It should not be set to `false`. Instead, it should be either `null` or
   * `false`. This is because using `null` means this key will not be stored in
   * Firebase and hence saving the storage needed for each question.
   */
  @Column({ nullable: true })
  preferNotToAnswer: true | null;

  /**
   * If `null` (and `preferNotToAnswer` is not true), it means that the user
   * clicks "Next" without answering.
   */
  @Column({ type: "simple-json", nullable: true })
  data: AnswerData | null;

  /**
   * The last update date for the answer (i.e., the last time the user interact
   * with this question - usually when the user clicks "Next").
   */
  @Column()
  date: Date;

  // https://github.com/microsoft/TypeScript/issues/5863#issuecomment-169173943
  static async findAnswerInPingAsync<T extends AnswerEntity>(
    ping: PingEntity,
    questionId: QuestionId,
  ): Promise<T | null> {
    const answer = await this.findOne({
      ping: { id: ping.id },
      questionId,
    });
    if (answer == null) {
      return null;
    }
    return answer as T;
  }
}

@ChildEntity(QuestionType.Slider)
export class SliderAnswerEntity extends AnswerEntity {
  data: SliderAnswerData | null;
}

@ChildEntity(QuestionType.ChoicesWithSingleAnswer)
export class ChoicesWithSingleAnswerAnswerEntity extends AnswerEntity {
  data: ChoicesWithSingleAnswerAnswerData | null;
}

@ChildEntity(QuestionType.ChoicesWithMultipleAnswers)
export class ChoicesWithMultipleAnswersAnswerEntity extends AnswerEntity {
  data: ChoicesWithMultipleAnswersAnswerData | null;
}

@ChildEntity(QuestionType.YesNo)
export class YesNoAnswerEntity extends AnswerEntity {
  data: YesNoAnswerData | null;
}

@ChildEntity(QuestionType.MultipleText)
export class MultipleTextAnswerEntity extends AnswerEntity {
  data: MultipleTextAnswerData | null;
}

@ChildEntity(QuestionType.HowLongAgo)
export class HowLongAgoAnswerEntity extends AnswerEntity {
  data: HowLongAgoAnswerData | null;
}

export function getAnswerEntity(questionType: QuestionTypeType): AnswerEntity {
  switch (questionType) {
    case QuestionType.Slider:
      return new SliderAnswerEntity();
    case QuestionType.ChoicesWithSingleAnswer:
      return new ChoicesWithSingleAnswerAnswerEntity();
    case QuestionType.ChoicesWithMultipleAnswers:
      return new ChoicesWithMultipleAnswersAnswerEntity();
    case QuestionType.YesNo:
      return new YesNoAnswerEntity();
    case QuestionType.MultipleText:
      return new MultipleTextAnswerEntity();
    case QuestionType.HowLongAgo:
      return new HowLongAgoAnswerEntity();
    default:
      throw new Error(`Unsupported question type ${questionType}`);
  }
}
