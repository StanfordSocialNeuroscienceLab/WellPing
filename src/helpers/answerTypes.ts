import { AnswerEntity } from "../entities/AnswerEntity";
import { Question, QuestionsList, Choice } from "./types";

// We use `{ value: ... }` because it seems that `"simple-json"`
// does not support simply mixing normal values.
// This also increases the consistency of the data.

export type SliderAnswerData = {
  value: number;
};

export type ChoicesWithSingleAnswerAnswerData = {
  value: Choice;
};

// We use an array of tuples here (instead of `{ [key: string]: boolean }`)
// because the order shown on the screen can be kept in the results.
// This is potentially helpful for analysis later if the order is randomized
// as now we know the order the user saw the choices in.
export type ChoicesWithMultipleAnswersAnswerChoices = [Choice, boolean][];
export type ChoicesWithMultipleAnswersAnswerData = {
  value: ChoicesWithMultipleAnswersAnswerChoices;
};

export type YesNoAnswerData = {
  value: boolean;
};

export type MultipleTextAnswerData = {
  value: string[];
};

export type HowLongAgoAnswerDataType = [number | null, string | null];
export type HowLongAgoAnswerData = {
  value: HowLongAgoAnswerDataType;
};

export type AnswerData =
  | SliderAnswerData
  | ChoicesWithSingleAnswerAnswerData
  | ChoicesWithMultipleAnswersAnswerData
  | YesNoAnswerData
  | MultipleTextAnswerData
  | HowLongAgoAnswerData;

export interface AnswersList {
  [id: string]: AnswerEntity;
}

export interface QuestionScreenProps {
  question: Question;
  onDataChange: (data: AnswerData) => void;
  allAnswers: AnswersList;
  allQuestions: QuestionsList;
  pipeInExtraMetaData: (input: string) => string;
  setDataValidationFunction: (func: () => boolean) => void;
}
