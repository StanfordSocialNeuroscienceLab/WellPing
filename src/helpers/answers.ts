import { AnswerData, Answer } from "./answerTypes";
import { getAnswersPingIdsQuestionIdsListAsync } from "./asyncStorage/answersPingIdsQuestionIdsList";
import { AnswerSchema } from "./schemas/Answer";
import {
  secureGetAnswerAsync,
  secureStoreAnswerAsync,
} from "./secureStore/answer";
import { Question, Ping } from "./types";

export async function getAnswersAsync(): Promise<Answer[]> {
  const answersList = await getAnswersPingIdsQuestionIdsListAsync();

  // https://stackoverflow.com/q/28066429/2603230
  const answers: Answer[] = await Promise.all(
    answersList.map(async (pingIdAndQuestionId) => {
      const answer = await secureGetAnswerAsync(
        pingIdAndQuestionId[0],
        pingIdAndQuestionId[1],
      );
      if (answer === null) {
        throw new Error(
          "answer === null in answersList.map in getAnswersAsync.",
        );
      }
      return answer;
    }),
  );

  return answers;
}

export async function insertAnswerAsync({
  ping,
  question,
  realQuestionId,
  preferNotToAnswer,
  data,
  date,
}: {
  ping: Ping;
  question: Question;
  realQuestionId: string;
  preferNotToAnswer: true | null; // See `MARK: WHY_PNA_TRUE_OR_NULL`.
  data: AnswerData | null;
  date: Date;
}): Promise<Answer> {
  const answer = AnswerSchema.parse({
    pingId: ping.id,
    questionId: realQuestionId,
    preferNotToAnswer,
    data,
    date,
  });

  await secureStoreAnswerAsync(answer);

  return answer;
}
