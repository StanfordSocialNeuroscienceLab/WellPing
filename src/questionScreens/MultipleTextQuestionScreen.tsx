import cloneDeep from "lodash/cloneDeep";
import React from "react";
import { View, TextInput, Alert } from "react-native";

import { MultipleTextAnswerEntity } from "../entities/AnswerEntity";
import {
  QuestionScreenProps,
  MultipleTextAnswerData,
} from "../helpers/answerTypes";
import { getNamesFileAsync } from "../helpers/studyFile";
import { MultipleTextQuestion, Names } from "../helpers/types";
// @ts-ignore
import SearchableDropdown from "../inc/react-native-searchable-dropdown";

interface MultipleTextQuestionScreenProps extends QuestionScreenProps {
  question: MultipleTextQuestion;
}

const MultipleTextQuestionScreen: React.ElementType<MultipleTextQuestionScreenProps> = ({
  question,
  onDataChange,
  allAnswers,
  setDataValidationFunction,
}) => {
  let numberOfTextFields = question.max;
  if (question.maxMinus) {
    const prevQuestionAnswer = allAnswers[
      question.maxMinus
    ] as MultipleTextAnswerEntity;
    if (prevQuestionAnswer && prevQuestionAnswer.data) {
      const length = prevQuestionAnswer.data.value.length;
      if (length > 0) {
        numberOfTextFields -= length;
      }
    }
  }

  const initTextValues = Array(numberOfTextFields).fill("");

  const [textValues, setTextValues] = React.useState<string[]>(initTextValues);

  React.useEffect(() => {
    // Re-fetch stored choices list (once) if necessary
    // TODO: ALLOW TO FIND LIST FROM DIFF. SOURCE STORED
    if (typeof question.choices == "string") {
      getNamesFileAsync().then((list) => {
        setPrestoredChoicesList(list);
      });
    }
  }, []);

  const updateTextValue = (text: string, index: number) => {
    const newTextValues: string[] = cloneDeep(textValues);
    newTextValues[index] = text.trim();
    setTextValues(newTextValues);

    const nonEmptyFields = newTextValues.filter(Boolean);
    const data: MultipleTextAnswerData = { value: nonEmptyFields };
    onDataChange(data);
  };

  const [prestoredChoicesList, setPrestoredChoicesList] = React.useState<
    string[]
  >([]);
  const prestoredChoicesListItems = prestoredChoicesList.map((name, index) => ({
    id: `${index}`,
    name,
  }));

  type DropdownItem = {
    id: string;
    name: string;
  };
  let textFieldsDropdownItems: DropdownItem[] = [];
  if (question.choices === "NAMES") {
    textFieldsDropdownItems = prestoredChoicesListItems;
  } else if (question.choices) {
    textFieldsDropdownItems = question.choices.map((choice) => ({
      id: choice.key,
      name: choice.value,
    }));
  }
  const textFieldsDropdownNames = textFieldsDropdownItems.map(
    (item) => item.name,
  );

  const textFields: SearchableDropdown[] = [];
  const textFieldsRef: React.RefObject<TextInput>[] = [];
  for (let index = 0; index < numberOfTextFields; index++) {
    const focusOnNextIfNotLast = () => {
      if (index !== numberOfTextFields - 1) {
        textFieldsRef[index + 1].current!.focus();
      }
    };

    textFieldsRef.push(React.useRef<TextInput>(null));

    textFields.push(
      <SearchableDropdown
        key={index}
        onItemSelect={(item: DropdownItem) => {
          updateTextValue(item.name, index);
          focusOnNextIfNotLast();
        }}
        containerStyle={{ padding: 5 }}
        itemStyle={{
          padding: 10,
          backgroundColor: "#F9F6EF",
          borderColor: "#bbb",
          borderWidth: 1,
        }}
        itemsContainerStyle={{ maxHeight: 100 }}
        items={textFieldsDropdownItems}
        resetValue={false}
        textInputProps={{
          ref: textFieldsRef[index],
          placeholder: question.placeholder,
          underlineColorAndroid: "transparent",
          style: {
            padding: 12,
            borderWidth: 1,
            borderColor: "#ccc",
            borderRadius: 5,
          },
          onChangeText: (text: string) => {
            updateTextValue(text, index);
          },
          onSubmitEditing: () => {
            focusOnNextIfNotLast();
          },
          onEndEditing: () => {
            dataValidationFunction(true);
          },
          accessibilityLabel: `text input ${index}`,
          accessibilityHint: "Enter your answer here",
        }}
        setSort={(item: DropdownItem, searchedText: string) =>
          // Match exact
          item.name.toLowerCase().startsWith(searchedText.toLowerCase())
        }
        listProps={{
          nestedScrollEnabled: true,
        }}
      />,
    );
  }

  let alertDisplaying = false;
  const dataValidationFunction = (
    showAlert?: boolean,
    values: string[] = textValues,
  ) => {
    if (!question.forceChoice) {
      // If there is no `forceChoice`, the data is always valid.
      return true;
    }

    if (alertDisplaying) {
      // It means the data is not valid, and an alert is still on screen.
      return false;
    }

    const invalidValues = values.filter((value) => {
      // If the text value is not empty and is not in the items list.
      return value && !textFieldsDropdownNames.includes(value);
    });

    if (invalidValues.length === 0) {
      // There is no invalid text values.
      return true;
    }

    if (showAlert) {
      alertDisplaying = true;

      const invalidValuesIndices = invalidValues.map((value) =>
        values.indexOf(value),
      );
      Alert.alert(
        "Notice",
        `You must select an item from the list.\n\n` +
          `The following ${
            invalidValues.length === 1 ? "item is" : "items are"
          } not in the list: ` +
          `${invalidValues.join(", ")}`,
        [
          {
            text: "OK",
            onPress: () => {
              invalidValuesIndices.forEach((index) => {
                textFieldsRef[index].current &&
                  textFieldsRef[index].current!.clear();
                // Because `clear()` does not call `onChangeText`.
                updateTextValue("", index);
              });
              textFieldsRef[invalidValuesIndices[0]].current!.focus();
              alertDisplaying = false;
            },
            style: "cancel",
          },
        ],
      );
    }
    return false;
  };
  setDataValidationFunction(() => {
    return dataValidationFunction(true);
  });

  return <View style={{ paddingTop: 10 }}>{textFields}</View>;
};

export default MultipleTextQuestionScreen;
