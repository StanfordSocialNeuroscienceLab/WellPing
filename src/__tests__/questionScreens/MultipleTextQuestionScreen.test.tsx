import React from "react";
import { Alert } from "react-native";
import {
  render,
  fireEvent,
  A11yAPI,
  FireEventAPI,
  waitFor,
} from "react-native-testing-library";
import { ReactTestInstance } from "react-test-renderer";

import { MultipleTextAnswerEntity } from "../../entities/AnswerEntity";
import { AnswersList, MultipleTextAnswerData } from "../../helpers/answerTypes";
import { QuestionType } from "../../helpers/helpers";
import { MultipleTextQuestion } from "../../helpers/types";
import MultipleTextQuestionScreen from "../../questionScreens/MultipleTextQuestionScreen";
import { simplePipeInExtraMetaData } from "../helper";

const A11Y_HINT = "Enter your answer here";
const getTextInputA11YLabel = (index: number) => `text input ${index}`;
const getTextInput = (
  index: number,
  getAllByA11yLabel: A11yAPI["getAllByA11yLabel"],
) => {
  const textInputs = getAllByA11yLabel(getTextInputA11YLabel(index));
  expect(textInputs).toHaveLength(1);
  const textInput = textInputs[0];
  return textInput;
};

const basicTestForQuestionAsync = async (
  question: MultipleTextQuestion,
  allAnswers: AnswersList,
  inputValues: string[],
) => {
  let codeDataValidationFunction: (() => boolean) | null = null;

  const mockOnDataChangeFn = jest.fn();
  const mockPipeInExtraMetaData = jest.fn(simplePipeInExtraMetaData);
  const mockSetDataValidationFunction = jest.fn((func) => {
    codeDataValidationFunction = func;
  });

  const renderResults = render(
    <MultipleTextQuestionScreen
      key={question.id}
      question={question}
      onDataChange={mockOnDataChangeFn}
      allAnswers={allAnswers}
      allQuestions={{ [question.id]: question }}
      pipeInExtraMetaData={mockPipeInExtraMetaData}
      setDataValidationFunction={mockSetDataValidationFunction}
    />,
  );
  const { getAllByA11yLabel, getAllByA11yHint } = renderResults;

  // Because there isn't a need to pipe in any data.
  expect(mockPipeInExtraMetaData).not.toHaveBeenCalled();

  expect(mockSetDataValidationFunction).toHaveBeenCalledTimes(1);
  expect(typeof codeDataValidationFunction).toBe("function");

  // This also helps test the placeholder property.
  const textInputs = getAllByA11yHint(A11Y_HINT);
  let textInputsLength = question.max;
  if (question.maxMinus) {
    if (allAnswers[question.maxMinus]) {
      const data = (allAnswers[question.maxMinus] as MultipleTextAnswerEntity)
        .data;
      textInputsLength -= (data ? data.value : []).length;
    }
  }
  expect(textInputs).toHaveLength(textInputsLength);
  expect(textInputs.length).toMatchSnapshot("text inputs length");

  const expectedAnswerData: MultipleTextAnswerData = { value: [] };
  let callCount = 0;
  for (let i = 0; i < textInputsLength; i++) {
    const textInput = getTextInput(i, getAllByA11yLabel);
    const inputValue = inputValues[i] || "";

    expect(textInput.props.placeholder).toBe(question.placeholder);

    fireEvent.changeText(textInput, inputValue);
    await waitFor(() => {
      const newTextInput = getTextInput(i, getAllByA11yLabel);
      return inputValue === newTextInput.props.value;
    });

    if (inputValue.length > 0) {
      expectedAnswerData.value[expectedAnswerData.value.length] = inputValue;
    }

    callCount += 1;
    expect(mockOnDataChangeFn).toHaveBeenNthCalledWith(
      callCount,
      expectedAnswerData,
    );
    expect(mockOnDataChangeFn).toHaveBeenCalledTimes(callCount);

    let isInputValid = true;
    if (question.choices && question.forceChoice && inputValue.length > 0) {
      let isInputInChoice = false;
      if (typeof question.choices === "string") {
        // TODO: DO THIS
      } else {
        isInputInChoice = question.choices?.includes(inputValue) || false;
      }

      if (!isInputInChoice) {
        let buttonPressed = false;

        const alertSpy = jest
          .spyOn(Alert, "alert")
          .mockImplementation((title, message, buttons) => {
            expect(message).toContain("You must select an item from the list");

            // Press "OK"
            buttons![0].onPress!();

            buttonPressed = true;
          });
        fireEvent(textInput, "onEndEditing", {});
        isInputValid = false;
        expect(alertSpy).toHaveBeenCalledTimes(1);

        await waitFor(() => buttonPressed);

        expectedAnswerData.value.pop();
        callCount += 1;
        expect(mockOnDataChangeFn).toHaveBeenNthCalledWith(
          callCount,
          expectedAnswerData,
        );

        alertSpy.mockRestore();
      }
    }
    if (isInputValid) {
      fireEvent(textInput, "onEndEditing", {});
      // There should be no extra call.
      expect(mockOnDataChangeFn).toHaveBeenNthCalledWith(
        callCount,
        expectedAnswerData,
      );
    }

    // TODO: test codeDataValidationFunction

    // TODO: it doesn't seems to be testing whether the text field (UI) is cleared.
  }

  // Store the expected data object.
  expect(expectedAnswerData).toMatchSnapshot();

  return renderResults;
};

const generateTypingInput = (length: number) => {
  return Array.from(Array(length), (_, i) => `I am typing ${i + 1}`);
};

test.each([
  [generateTypingInput(1), 3, "Enter something...", "INDEX"],
  [generateTypingInput(4), 4, "", "ITEM"],
  [["John Doe", "王小明"], 5, "Enter a name...", "NAMEINDEX"],
  [
    ["Mr. Fox", "Felicity Fox", "Ash Fox", "Kristofferson Silverfox"],
    4,
    undefined,
    "FOX_INDEX",
  ],
])(
  "input `%p` with max %d without choices",
  async (inputValues, max, placeholder, indexName) => {
    const question = {
      id: "WithoutChoicesDict",
      type: QuestionType.MultipleText,
      question: "A question",
      placeholder,
      max,
      variableName: "TARGET_CATEGORY",
      indexName,
      next: null,
    } as MultipleTextQuestion;

    await basicTestForQuestionAsync(question, {}, inputValues);
  },
);

test.each([
  [
    generateTypingInput(1),
    3,
    "PrevQuestionId",
    "Enter something...",
    "INDEX",
    {},
  ],
  [
    [],
    4,
    "PrevQuestionId",
    "Enter something...",
    "INDEX",
    {
      PrevQuestionId: {
        questionId: "PrevQuestionId",
        data: {
          value: [],
        },
      },
    },
  ],
  [
    ["John Doe", "王小明"],
    4,
    "FamilyNames",
    "Enter a name...",
    "NAMEINDEX",
    {
      FamilyNames: {
        questionId: "FamilyNames",
        data: {
          value: ["Father", "Mother"],
        },
      },
    },
  ],
  [
    ["Kristofferson Silverfox"],
    4,
    "ZootopiaCharacters",
    undefined,
    "FOX_INDEX",
    {
      ZootopiaCharacters: {
        questionId: "ZootopiaCharacters",
        data: {
          value: ["Judy Hopps", "Nick Wilde", "Flash"],
        },
      },
    },
  ],
])(
  "input `%p` with max %d and maxMinus %p without choices",
  async (inputValues, max, maxMinus, placeholder, indexName, allAnswers) => {
    const question = {
      id: "WithoutChoicesWithMaxMinusDict",
      type: QuestionType.MultipleText,
      question: "A question",
      placeholder,
      max,
      maxMinus,
      variableName: "TARGET_CATEGORY",
      indexName,
      next: null,
    } as MultipleTextQuestion;

    await basicTestForQuestionAsync(question, allAnswers as any, inputValues);
  },
);

const CHOICES = [
  "Friend",
  "Co-worker",
  "Parent",
  "Sibling / other relative",
  "Significant other",
  "Stranger",
  "Other",
];
test.each([
  [generateTypingInput(2), 2, true, "Enter a relation..."],
  [
    [
      ...generateTypingInput(1),
      "Sibling / other relative",
      ...generateTypingInput(2),
      "Parent",
    ],
    5,
    true,
    "Enter a relation...",
  ],
  [generateTypingInput(4), 4, false, undefined],
  [["Friend", "RANDOM INPUT", "Other"], 5, true, "Enter a relation..."],
  [
    ["Stranger", "Stranger", "Stranger", "Stranger", "RANDOM"],
    5,
    false,
    undefined,
  ],
  [["Co-worker", "", "Co-worker", ":)", "Co-worker"], 5, true, "Relation..."],
])(
  "input `%p` with max %d and forceChoice %p with a choices object",
  async (inputValues, max, forceChoice, placeholder) => {
    const question = {
      id: "WithChoicesDict",
      type: QuestionType.MultipleText,
      question: "A question",
      placeholder,
      choices: CHOICES,
      forceChoice,
      max,
      variableName: "TARGET_CATEGORY",
      indexName: "INDEX",
      next: null,
    } as MultipleTextQuestion;

    await basicTestForQuestionAsync(question, {}, inputValues);
  },
);

// TODO: TEST CHOICES WITH EXTERNAL JSON
