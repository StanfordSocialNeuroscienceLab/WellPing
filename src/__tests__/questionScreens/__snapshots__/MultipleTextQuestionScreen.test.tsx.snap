// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`input \`["Co-worker", "", "Co-worker", ":)", "Co-worker"]\` with max 5 and forceChoice true with a choices object 1`] = `
Object {
  "value": Object {
    "Input_1": "Co-worker",
    "Input_2": "Co-worker",
    "Input_3": "Co-worker",
  },
}
`;

exports[`input \`["Friend", "RANDOM INPUT", "Other"]\` with max 5 and forceChoice true with a choices object 1`] = `
Object {
  "value": Object {
    "Input_1": "Friend",
    "Input_2": "Other",
  },
}
`;

exports[`input \`["I am typing 1", "I am typing 2", "I am typing 3", "I am typing 4"]\` with max 4 and forceChoice false with a choices object 1`] = `
Object {
  "value": Object {
    "Input_1": "I am typing 1",
    "Input_2": "I am typing 2",
    "Input_3": "I am typing 3",
    "Input_4": "I am typing 4",
  },
}
`;

exports[`input \`["I am typing 1", "I am typing 2", "I am typing 3", "I am typing 4"]\` with max 4 without choices 1`] = `
Object {
  "value": Object {
    "1": "I am typing 1",
    "2": "I am typing 2",
    "3": "I am typing 3",
    "4": "I am typing 4",
  },
}
`;

exports[`input \`["I am typing 1", "I am typing 2"]\` with max 2 and forceChoice true with a choices object 1`] = `
Object {
  "value": Object {},
}
`;

exports[`input \`["I am typing 1", "Sibling / other relative", "I am typing 1", "I am typing 2", "Parent"]\` with max 5 and forceChoice true with a choices object 1`] = `
Object {
  "value": Object {
    "Input_1": "Sibling / other relative",
    "Input_2": "Parent",
  },
}
`;

exports[`input \`["I am typing 1"]\` with max 3 without choices 1`] = `
Object {
  "value": Object {
    "Typing_1": "I am typing 1",
  },
}
`;

exports[`input \`["John Doe", "王小明"]\` with max 5 without choices 1`] = `
Object {
  "value": Object {
    "Name_1": "John Doe",
    "Name_2": "王小明",
  },
}
`;

exports[`input \`["Mr. Fox", "Felicity Fox", "Ash Fox", "Kristofferson Silverfox"]\` with max 4 without choices 1`] = `
Object {
  "value": Object {
    "Fox1": "Mr. Fox",
    "Fox2": "Felicity Fox",
    "Fox3": "Ash Fox",
    "Fox4": "Kristofferson Silverfox",
  },
}
`;

exports[`input \`["Stranger", "Stranger", "Stranger", "Stranger", "RANDOM"]\` with max 5 and forceChoice false with a choices object 1`] = `
Object {
  "value": Object {
    "Input_1": "Stranger",
    "Input_2": "Stranger",
    "Input_3": "Stranger",
    "Input_4": "Stranger",
    "Input_5": "RANDOM",
  },
}
`;
